lib/execution_segment.go

```diff
@@ -394,13 +396,275 @@ func (ess ExecutionSegmentSequence) String() string {
+// denominator. So 3/5 becomes 9/15 and 1/3 becomes 5/15. So now for each 15 elements 9 need to go
+// to the 3/5, and 5 need to go to 1/3.
+//
+// We use the below algorithm to split elements between ExecutionSegments by using their length as
+// the rational number. As we would like to get non sequential elements we try to get the maximum
+// distance between them. That is the number of elements divided by the number of elements for any
+// given segment, which concidently is the length of the segment reversed.
+// The algorithm below does the following:
+// 1. Goes through the elements from 0 to the lcd-1
+// 2. For each of element goes through the segments and looks if the amount of already taken
+// elements by the given segment multiplied by that segment length inverted is equal to or less to
+// the current element index. if it is give that element to that segment if not continue with the
+// next element.
 //
-// TODO: add a more detailed algorithm description
-func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int, []int, error) {
-	start := 0
-	offsets := []int{}
-	// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
-	return start, offsets, fmt.Errorf("not implemented")
+// The code below specifically avoids using big.Rat which complicates the code somewhat.
+// As additional note the sorting of the segments from biggest to smallest helps with the fact that
+// the biggest elements will need to take the most elements and for them it will be the hardest to
+// not get sequential elements.
+func (e sortInterfaceWrapper) stripingAlgorithm(saveIndex func(iteration int64, index int, numerator int64) bool) {
+	var chosenCounts = make([]int64, len(e.slice))
+
+outer:
+	for i := int64(0); i < e.lcd; i++ {
+		for index, chosenCount := range chosenCounts {
+			num := chosenCount * e.lcd
+			denom := e.slice[index].numerator
+			if i > num/denom || (i == num/denom && num%denom == 0) {
+				chosenCounts[index]++
+				if saveIndex(i, e.slice[index].originalIndex, denom) {
+					break outer
+				}
+				break
+			}
+		}
+	}
+}
+
+// ExecutionTuple is here to represent the combination of ExecutionSegmentSequence and
+// ExecutionSegment and to give easy access to a couple of algorithms based on them in a way that is
+// somewhat perfomant for which it generally needs to cache the results
+type ExecutionTuple struct { // TODO rename
+	ES *ExecutionSegment // TODO unexport this as well?
+
+	// TODO: have the index of the segment, cached?
+	sequence     ExecutionSegmentSequence
+	offsetsCache [][]int64
+	lcd          int64
+	// TODO discuss if we just don't want to fillCache in the constructor and not need to use pointer receivers everywhere
+	once *sync.Once
+}
+
+func fillSequence(sequence ExecutionSegmentSequence) ExecutionSegmentSequence {
+	// TODO: discuss if we want to get the lcd of the sequence and fill with it elements of length 1/lcd ?
```

Comment:
it is certainly the most straightforward one to explain as well ... because all the others get hairy with strange segments such as "1/3:2/5" or something like that ... 

The biggest problem in my experience is that if you run it with "0:1/3", "1/3:2/3" and "2/3:1". the first and last behave the same while the middle one behaves as I have started to expect (which is not to say that it is wrong). 
I don't think it is noticeable if you don't look for it and for the majority of the use cases I would argue this will work just the same :D

Reply:
> The biggest problem in my experience is that if you run it with "0:1/3", "1/3:2/3" and "2/3:1". the first and last behave the same while the middle one behaves as I have started to expect (which is not to say that it is wrong).

But that's the thing - if you fill in the sequence with `1/lcd` pieces, you'd get this nasty behavior pretty much all the time - equally sized segments would behave absolutely identically. So your current solution would be a better fit for them, I think... In any case, you probably should delete the `TODO` from here, to avoid confusion - we're not going to change the current implementation without some very good reason.